aten/src/ATen/cpu/vec256/vec256_bfloat16.h

```diff
@@ -428,30 +427,28 @@ Vec256<BFloat16> inline operator^(const Vec256<BFloat16>& a, const Vec256<BFloat
   return _mm256_xor_si256(a, b);
 }
 
-const Vec256<BFloat16> Vec256<BFloat16>::ones(1.0f);
```

Comment:
Can we create these constants statically inside each function that uses it? This should speed things up (no need to do the initialization step every time `ones` is needed) and it seems like they won't initialize until that function is executed for the first time: <https://stackoverflow.com/a/58804/1150462>

> An implementation is permitted to perform early initialization of other local objects with static storage duration under the same conditions that an implementation is permitted to statically initialize an object with static storage duration in namespace scope (basic.start.init). **Otherwise such an object is initialized the first time control passes through its declaration**; such an object is considered initialized upon the completion of its initialization.

Reply:
No, please do not do this.

1) This is not an effective optimization. GCC/clang are already really good at lifting out these sorts of constants from inner loops.
2) It introduces atomic guards for the initialization. GCC/clang are not good at optimizing away the guards or lifting them out of loops.

Reply:
I now agree with Point 1 (after some experimentation). Mind explaining Point 2 a bit?

Reply:
Here's an illustrative example. I've used `-Os` to make the assembly easier to read, but the same issue exists with `-O3`:

`and_ones_v1`: re-creates the "one" variable every time
`and_ones_v2`: uses function-local static data for the "one" variable

https://gcc.godbolt.org/z/gh4gps

Non-trivial static data needs to be initialized exactly once in a thread-safe manner. The compiler does this by adding a guard variable and a mutex. If the guard is set, the static data was already initialized. If it's not set, then it acquires the mutex, initializes the data and sets the guard variable. This is efficient for most C++ code, but can be slow for tight inner loops like in our kernels.

Current compilers aren't good at lifting the guard variable check outside the inner loop, so you end with at least an extra memory load and test+jmp in the inner loop. The generated call also seems to pessimize with other compiler analyses so the "one" variable is still reloaded every time through the loop.

In the hot-path through the loop `and_ones_v1` has one vector load (implicit in vandps) and one vector store. In the hot-path through the loop `and_ones_v2` has an extra vector load with `-O3` and two extra vector loads and an extra store in `-Os`.

Reply:
Thanks for the explanation, this is really helpful.